### ISSUE TITLE
no-use-before-define should have option to look at scope

### ISSUE THREAD

**Issue by Thaina** [eslint/eslint#6499]
no-use-before-define should have some option to allow define in outer scope after inner function

```js
{
    "no-use-before-define" : "onlyScopeError"
}


function A()
{
    var c = config; // should not error
    value = 0; // should error

    var value;
}

var config = 0;
```

**Comment by mysticatea** [comment]
Thank you for this issue.

ESLint is a static analyzer, so it cannot disallow unexpected uses of outer scope's variable correctly.
`no-use-before-define` rule is merely checking locations between a use of a variable and the definition of the variable.

```js
function a() {
    console.log(b) // Is b initialized or uninitialized? -- It's unknown in static analysis.
}

foo(a) // might be `function foo(f) { f() }`

var b = 1
```

I'm not sure the option to ignore a use of outer scope's variables is good one. The option would introduce overlooking of mistakes.

**Comment by ilyavolodin** [comment]
Introduction of such option would basically be the same as turning the rule off completely. This rule does not exist to flag invalid JavaScript code, but to allow developers that are used to other languages that do not have hoisting to enforce their preferences. The code you provided above will execute correctly and is a valid JavaScript code, but goes against preferences of some of the developers.

As @mysticatea mentioned, this is not something that we can do through static analysis easily. Closing as this is not something we can implement correctly.

**Comment by skeggse** [comment]
I'd like to petition to reopen this issue, with a slightly restricted scope (sigh).

In node, it's an antipattern to do much more than initialization at the top-level module scope. It's also common to see a flow from high-level declarations to low-level ones. You might see a high-level dispatch function followed by a hash of lower-level handlers. It would be really nice to be able to disable this rule specifically for references to bindings defined at the top-level scope (relative to a module or file).

The example given by @mysticatea is correct, but in some contexts not significant; developers enabling this option should understand what they're opting in to, and documentation can help with that.
